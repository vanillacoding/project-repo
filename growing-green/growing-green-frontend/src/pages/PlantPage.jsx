import React, { useEffect } from 'react';
import styled from 'styled-components';
import { useSelector, useDispatch } from 'react-redux';
import { useParams, useHistory } from 'react-router-dom';
import { getAllPlantsByUserId } from '../redux/modules/plants';

import WeatherInfo from '../components/WeatherInfo';
import Calendar from '../components/Calendar';
import PlantCanvas from '../components/PlantCanvas';
import TimeTravelCanvas from '../components/TimeTravelCanvas';
import ErrorBox from '../components/ErrorBox';
import Loading from '../components/Loading';
import TimeTravelMode from '../components/TimeTravelMode';

import newPlantIcon from '../assets/images/background/new_plant_icon.png';
import backButton from '../assets/images/arrows/back_arrow.png';

export default function PlantPage() {
  const { allPlants, isLoading, error } = useSelector((state) => state.plants);
  const { isTimeTravelMode, plantsInWeek } = useSelector(
    (state) => state.timeTravel,
  );

  const { plantId } = useParams();
  const dispatch = useDispatch();
  const history = useHistory();

  const plantIds = Object.keys(allPlants);
  const currentIndex = plantIds.indexOf(plantId);
  const plants = Object.values(allPlants);

  useEffect(() => {
    dispatch(getAllPlantsByUserId());
  }, [dispatch, plantId]);

  function renderError() {
    return <ErrorBox message={error} />;
  }

  function renderLoading() {
    return <Loading text="식물정보 불러오는중..." />;
  }

  function renderPage() {
    return (
      <>
        <TimeTravelMode />
        {isTimeTravelMode === true ? (
          <TimeTravelCanvas
            plantsInWeek={plantsInWeek}
            isTimeTravelMode={isTimeTravelMode}
          />
        ) : (
          <>
            <CalendarAndWeather>
              <Calendar />
              <WeatherInfo height="120" temperature="30" icon />
            </CalendarAndWeather>
            <PlantCanvas plants={plants} isTimeTravelMode={isTimeTravelMode} />
            <NewPlantButtonWrapper>
              <NewPlantButton onClick={() => history.push('/create')} />
            </NewPlantButtonWrapper>
            <BackButton onClick={() => history.push('/')} />
          </>
        )}
      </>
    );
  }

  return (
    <Container>
      {error ? renderError() : isLoading ? renderLoading() : renderPage()}
    </Container>
  );
}

const Container = styled.div`
  position: relative;
  width: 1200px;
  height: 700px;
`;

const CalendarAndWeather = styled.div`
  position: absolute;
  align-items: center;
  left: 4rem;
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  align-self: flex-start;
`;

const LeftArrow = styled.img`
  position: absolute;
  left: 8rem;
  top: 14rem;
  width: 150px;
`;

const RightArrow = styled.img`
  position: absolute;
  right: 8rem;
  top: 14rem;
  width: 150px;
`;

const NewPlantButtonWrapper = styled.div`
  position: absolute;
  right: 4.5rem;
  top: 6rem;
  width: 115px;
  height: 115px;
  background: #a9cf98;
  border-radius: 100px;
  box-shadow: rgba(0, 0, 0, 0.4) 0px 2px 4px,
    rgba(0, 0, 0, 0.3) 0px 7px 13px -3px,
    rgba(20, 20, 20, 0.2) 3px -3px 0px inset;
`;

const NewPlantButton = styled.button`
  background: url(${newPlantIcon});
  background-size: cover;
  position: absolute;
  right: -14px;
  top: -14px;
  width: 150px;
  height: 170px;
  border: none;
`;

const BackButton = styled.button`
  width: 74px;
  height: 50px;
  position: absolute;
  left: 2.5rem;
  border: none;
  bottom: 1rem;
  background: url(${backButton}) no-repeat;
  background-size: cover;
`;
