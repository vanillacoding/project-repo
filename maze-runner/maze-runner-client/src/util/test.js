import { NODE_STATUS } from '../constant';
import { createNodes } from './maze';

export function getNodeColIndex(nodeId) {
  if (typeof nodeId !== 'string' && !(nodeId instanceof String)) {
    throw new Error('Node id should be a string');
  }

  if (nodeId.split('-').length !== 2) {
    throw new Error('Invalid node id');
  }

  return nodeId.split('-')[1];
}

export function getNodeRowIndex(nodeId) {
  if (typeof nodeId !== 'string' && !(nodeId instanceof String)) {
    throw new Error('Node id should be a string');
  }

  if (nodeId.split('-').length !== 2) {
    throw new Error('Invalid node id');
  }

  return nodeId.split('-')[0];
}

export function getPrevNodeId(nodeId) {
  if (typeof nodeId !== 'string' && !(nodeId instanceof String)) {
    throw new Error('Node id should be a string');
  }

  if (nodeId.split('-').length < 2) {
    throw new Error('Invalid node id');
  }

  const row = getNodeRowIndex(nodeId);
  const col = getNodeColIndex(nodeId);

  return `${Number(row)}-${Number(col) - 1}`;
}

export function createById(startNodeId, endNodeId) {
  const exampleBlocks = [
    [0, 0, 0, 1, 0, 0, 1, 0, 0, 1],
    [0, 1, 0, 1, 0, 1, 1, 0, 1, 1],
    [0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 1, 0, 1, 1, 1, 1, 1, 1, 1],
    [0, 0, 0, 1, 0, 1, 1, 1, 0, 0],
    [0, 1, 1, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 1, 1, 1, 1, 1, 0],
    [1, 1, 1, 0, 1, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 1, 0, 1, 1, 0, 1],
    [1, 0, 1, 0, 0, 0, 0, 1, 0, 0],
  ];

  const startNodeIndex = {
    col: getNodeColIndex(startNodeId),
    row: getNodeRowIndex(startNodeId),
  };
  const endNodeIndex = {
    col: getNodeColIndex(endNodeId),
    row: getNodeRowIndex(endNodeId),
  };

  exampleBlocks[startNodeIndex.row][startNodeIndex.col] = 3;
  exampleBlocks[endNodeIndex.row][endNodeIndex.col] = 4;

  const [widthCount, heightCount] = [
    exampleBlocks[0].length,
    exampleBlocks.length,
  ];

  const nodes = createNodes(widthCount, heightCount);
  const { byId } = nodes;

  exampleBlocks.forEach((row, rowIndex) =>
    row.forEach((number, colIndex) => {
      const nodeId = `${rowIndex}-${colIndex}`;

      switch (number) {
        case 0: {
          byId[nodeId].status = NODE_STATUS.UNVISITED;
          return;
        }
        case 1: {
          byId[nodeId].status = NODE_STATUS.WALL;
          return;
        }
        case 2: {
          byId[nodeId].status = NODE_STATUS.WEIGHTED;
          byId[nodeId].weight = 10;
          return;
        }
        case 3: {
          byId[nodeId].status = NODE_STATUS.START;
          return;
        }
        case 4: {
          byId[nodeId].status = NODE_STATUS.END;
          return;
        }
        case 5: {
          byId[nodeId].status = NODE_STATUS.MIDDLE;
          return;
        }
        default: {
          return;
        }
      }
    }),
  );

  return byId;
}

export default {
  getNodeColIndex,
  getNodeRowIndex,
  getPrevNodeId,
  createById,
};
