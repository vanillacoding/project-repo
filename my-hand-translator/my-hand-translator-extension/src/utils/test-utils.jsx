import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";

import userReducer from "../features/user/userSlice";

function render(
  ui,
  {
    preloadedState,
    store = configureStore({ reducer: { user: userReducer }, preloadedState }),
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from "@testing-library/react";

const user = {
  projectId: "translate-324702",
  translations: [
    {
      nanoId: "qWYBIfLPMsjP1LHNRqlFW",
      origin: "react",
      translated: "리액트",
      url: "vanilla",
      glossary: {
        react: "리액트",
      },
      createdAt: "2017-01-01",
    },
  ],
  glossary: {
    react: "리액트",
  },
  glossaryId: "6141bd1c2a11e5b51320bac1",
  isServerOn: false,
  name: "aidencoders",
  email: "aidencoders@gmail.com",
};

export { render, user };
