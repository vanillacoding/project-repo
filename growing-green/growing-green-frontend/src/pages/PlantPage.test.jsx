import React from 'react';
import { render, fireEvent, waitFor, cleanup } from '../test/test-utils';
import PlantPage from './PlantPage';

describe('<PlantPage>', () => {
  afterEach(() => cleanup());

  it('matches snapshot', () => {
    const { container } = render(<PlantPage />);
    expect(container).toMatchSnapshot();
  });

  it('should load and render plant canvas after fetch all plant', async () => {
    const { getByTestId, getByText } = render(<PlantPage />);
    expect(getByText('식물정보 불러오는중...')).toBeInTheDocument();

    await waitFor(() => getByTestId('canvas'));

    expect(getByTestId('canvas')).toBeInTheDocument();
    expect(getByText('mockWeatherName')).toBeInTheDocument();
  });

  it('should render time traveling canvas after clicking start time traveling button', async () => {
    const { getByTestId, getByText, getByRole } = render(<PlantPage />);
    await waitFor(() => getByTestId('canvas'));

    fireEvent.click(getByRole('button', { name: 'Start time traveling' }));
    expect(getByText('Stop time traveling')).toBeInTheDocument();
    expect(getByTestId('travelCanvas')).toBeInTheDocument();
  });
});
