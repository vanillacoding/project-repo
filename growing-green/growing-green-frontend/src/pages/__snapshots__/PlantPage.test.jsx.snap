// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlantPage> matches snapshot 1`] = `
<div>
  <div
    class="sc-fotOHu kOEZrp"
  >
    <div
      class="sc-kDTinF kxZmHi"
    >
      <div
        class="sc-iqseJM cfpYBU"
        data-testid="loading-spinner"
      />
      <p>
        식물정보 불러오는중...
      </p>
    </div>
  </div>
</div>
`;
